### PR TITLE
Prevent dangling GetAttribute calls

### LINKED ISSUE #18734: Git.exe Processes Hanging / Not Cleaning Up
### Gitea Version

1.16.1

### Git Version

2.35.1.windows.2

### Operating System

Windows Server 2012 R2

### How are you running Gitea?

 - Using the prebuilt package for 1.16.1 - gitea-1.16.1-windows-4.0-amd64.exe
 - Running via IIS with a forward proxy to pass traffic to the Gitea instance.
 - Gitea is running as a system service.

### Database

MySQL

### Can you reproduce the bug on the Gitea demo site?

No

### Log Gist

_No response_

### Description

When running for an extended period of time, we are noticing that git.exe instances are left open/hanging. After about 25-30 instances are left open, the site just starts throwing 500 errors. It becomes impossible to do anything on the site outside of viewing the main page and admin panel.

This issue seems to be an exact copy of this: https://github.com/go-gitea/gitea/issues/17138

I am seeing the same thing.

Monitor page shows tons of entries for things like this:

```
C:\Program Files\Git\cmd\git.exe -c credential.helper= -c protocol.version=2 -c uploadpack.allowfilter=true check-attr --stdin -z --cached linguist-vendored linguist-generated linguist-language gitlab-language -- [repo_path: PATH\TO\REPO\HERE.git]
```

There are also other monitored processes stuck:

```
GET: /PATH/TO/REPO/HERE/commits/commit/COMMIT_HASH_HERE/path/to/file.h?lang=pt-BR
3 hours ago
  C:\Program Files\Git\cmd\git.exe cat-file --batch-check [repo_path: PATH\TO\REPO\HERE.git] (modules/context/repo.go:559)
  3 hours ago
  C:\Program Files\Git\cmd\git.exe cat-file --batch [repo_path: PATH\TO\REPO\HERE.git] (modules/context/repo.go:559)
  3 hours ago
```

```
GET: /PATH/TO/REPO/HERE/commit/COMMIT_HASH_HERE.diff
2 hours ago
  C:\Program Files\Git\cmd\git.exe cat-file --batch [repo_path: PATH\TO\REPO\HERE.git] (modules/context/repo.go:559)
  2 hours ago
  C:\Program Files\Git\cmd\git.exe cat-file --batch-check [repo_path: PATH\TO\REPO\HERE.git] (modules/context/repo.go:559)
  2 hours ago
```

```
C:\Program Files\Git\cmd\git.exe cat-file --batch [repo_path: PATH\TO\REPO\HERE.git] (modules/git/repo_base_nogogit.go:41)
```

And so on. The main one is the first reported entry, there are lots of those ones.

This is a bit rough to diagnose as it does not happen immediately. It takes a few hours before we start seeing the instances begin hanging. 

### Screenshots

_No response_

### PR DESCRIPTION
**Pull request by zeripath**
It appears possible that there could be a hang due to unread data from the
repo-attribute command pipes. This PR simply closes these during the defer.

Fix #18734

Signed-off-by: Andrew Thornton <art27@cantab.net>

### CODE REVIEW

**Review by delvh — APPROVED** [review]

**Comment by lunny** [comment]
CI failed is related.

**Review by 6543 — APPROVED** [review]

**Comment by 6543** [comment]
**DATA RACE** -- hmm do we need some kind of rw lock ?

File: services/gitdiff/gitdiff.go [hunk 1]
```diff
@@ -1418,10 +1418,12 @@ func GetDiff(gitRepo *git.Repository, opts *DiffOptions, files ...string) (*Diff
 					if err != nil && err != ctx.Err() {
 						log.Error("Unable to open checker for %s. Error: %v", opts.AfterCommitID, err)
 					}
+					_ = checker.Close()
 					cancel()
 				}()
 			}
 			defer func() {
+				_ = checker.Close()
```
Comment (6543) [inline services/gitdiff/gitdiff.go:1425]: I think it should be fine to close only outside the go call ...

**Comment by 6543** [comment]
:rocket:

**Comment by zeripath** [comment]
Backport #18755

**Pull request merged**